File: frontend/types/organization.ts
/**
 * Foundational Organization, Membership, and RBAC Permission Types.
 * Matches the OfficeOS multi-tenant architecture and FastAPI backend schemas.
 */

export type CanonicalRole =
  | "system_admin"
  | "organization_owner"
  | "organization_admin"
  | "hr_manager"
  | "finance_manager"
  | "project_manager"
  | "department_manager"
  | "employee";

export type UserRole = CanonicalRole | string;

export type CanonicalPermission =
  | "organizations.view"
  | "organizations.update"
  | "organizations.settings_manage"
  | "branches.view"
  | "branches.manage"
  | "departments.view"
  | "departments.manage"
  | "roles.view"
  | "members.view"
  | "members.manage"
  | "employees.view"
  | "employees.create"
  | "employees.update"
  | "employees.delete"
  | "org:read"
  | "org:update"
  | "org:delete"
  | "employee:read"
  | "employee:create"
  | "employee:update"
  | "employee:delete"
  | "attendance:read"
  | "attendance:create"
  | "attendance:approve"
  | "leave:read"
  | "leave:create"
  | "leave:approve"
  | "project:read"
  | "project:create"
  | "project:update"
  | "finance:read"
  | "finance:manage"
  | "settings:read"
  | "settings:manage";

export type Permission = CanonicalPermission | string;

export interface PermissionData {
  code: string;
}

export interface Organization {
  id: string;
  name: string;
  slug: string;
  logoUrl?: string;
  createdAt?: string;
  updatedAt?: string;
}

export interface OrganizationProfile {
  id: string;
  name: string;
  slug: string;
  is_active: boolean;
  created_at: string;
  updated_at: string;
}

export interface OrganizationProfileUpdate {
  name?: string;
}

export interface OrganizationSettings {
  id: string;
  organization_id: string;
  timezone: string;
  currency: string;
  created_at: string;
  updated_at: string;
}

export interface OrganizationSettingsUpdate {
  timezone?: string;
  currency?: string;
}

export interface BranchResponse {
  id: string;
  organization_id: string;
  name: string;
  code: string;
  address?: string | null;
  is_active: boolean;
  created_at: string;
  updated_at: string;
}

export type Branch = BranchResponse;

export interface BranchCreate {
  name: string;
  code: string;
  address?: string | null;
  is_active?: boolean;
}

export interface BranchUpdate {
  name?: string;
  code?: string;
  address?: string | null;
  is_active?: boolean;
}

export interface RoleResponse {
  id: string;
  organization_id?: string | null;
  name: string;
  description?: string | null;
  is_system: boolean;
}

export interface MemberProfile {
  id: string;
  email?: string | null;
  first_name?: string | null;
  last_name?: string | null;
}

export interface MemberResponse {
  id: string;
  organization_id: string;
  profile_id: string;
  status: string;
  profile: MemberProfile;
  roles: RoleResponse[];
  created_at: string;
  updated_at: string;
}

export interface OrganizationMembership {
  id: string;
  organizationId: string;
  organization: Organization;
  userId: string;
  role?: UserRole | null;
  permissions: Permission[];
  isActive: boolean;
  createdAt: string;
}

File: backend/app/schemas/organization.py
from datetime import datetime
from typing import Any
from uuid import UUID

from pydantic import BaseModel, ConfigDict, Field


class OrganizationProfileResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: UUID
    name: str
    slug: str
    is_active: bool
    created_at: datetime
    updated_at: datetime


class OrganizationProfileUpdate(BaseModel):
    name: str | None = Field(None, min_length=2, max_length=160)


class OrganizationSettingsResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: UUID
    organization_id: UUID
    timezone: str
    currency: str
    created_at: datetime
    updated_at: datetime


class OrganizationSettingsUpdate(BaseModel):
    timezone: str | None = Field(None, min_length=1, max_length=64)
    currency: str | None = Field(None, min_length=3, max_length=3)


class BranchResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: UUID
    organization_id: UUID
    name: str
    code: str
    address: str | None = None
    is_active: bool
    created_at: datetime
    updated_at: datetime


class BranchCreate(BaseModel):
    name: str = Field(..., min_length=2, max_length=160)
    code: str = Field(..., min_length=1, max_length=32)
    address: str | None = Field(None, max_length=500)
    is_active: bool = True


class BranchUpdate(BaseModel):
    name: str | None = Field(None, min_length=2, max_length=160)
    code: str | None = Field(None, min_length=1, max_length=32)
    address: str | None = Field(None, max_length=500)
    is_active: bool | None = None


class RoleResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: UUID
    organization_id: UUID | None = None
    name: str
    description: str | None = None
    is_system: bool


class MemberProfile(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: UUID
    email: str | None = None
    first_name: str | None = None
    last_name: str | None = None


class MemberResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: UUID
    organization_id: UUID
    profile_id: UUID
    status: str
    profile: MemberProfile
    roles: list[RoleResponse] = Field(default_factory=list)
    created_at: datetime
    updated_at: datetime


class MemberAddRequest(BaseModel):
    email: str = Field(..., min_length=3, max_length=320)
    role_ids: list[UUID] = Field(default_factory=list)
    status: str = Field("active", pattern="^(active|invited|suspended)$")


class MemberUpdateRequest(BaseModel):
    role_ids: list[UUID] | None = None
    status: str | None = Field(None, pattern="^(active|invited|suspended)$")


class AuditLogResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: UUID
    organization_id: UUID
    actor_id: UUID | None = None
    actor_email: str | None = None
    action: str
    entity_type: str
    entity_id: UUID | None = None
    details: dict[str, Any] | None = None
    ip_address: str | None = None
    created_at: datetime
